attrs_strict/_type_validation.py

```diff
@@ -16,7 +16,7 @@ def type_validator(empty_ok=True):
 
     def _validator(instance, attribute, field):
         if not empty_ok and not field:
-            raise ValueError("Expected a non-empty attribute")
+            raise EmptyError(field, attribute)
```

Comment:
Could you please write some test for this, eventually including some cases where we have a recursive type (see other tests).

Reply:
I changed test case 'test_raises_when_container_is_empty_and_empty_ok_is_false' to match with my proposal. This test also proves that you can catch the exception by 'except ValueError.'

What do you mean by 'recursive type'? If this isn't enough as test code, could you tell me more precisely?

Reply:
So I was thinking something like: 

```python
@attr.s
class Something():
   name = attr.ib(type=Dict[str, List[str]], validator=type_validator)
```
And after that:
```python
smth = Something(name={"foo": []})
``` 
I think this should trigger it, basically you have a list inside a dictionary, and the list should NOT be empty.

Reply:
I tried the following test code but failed because it didn't raise error.

```python
def test_raises_when_nested_value_is_empty():
    # GIVEN
    @attr.s
    class Something():
        name = attr.ib(type=typing.Dict[str, typing.List[str]],
                       validator=type_validator(empty_ok=False))

    # WHEN
    with pytest.raises(ValueError) as error:
        Something(name={"foo": []})

    # THEN (TODO: should be exact equal check)
    assert "" != str(error.value)

```

So I went back to the original master branch(without my proposal patch), but I got no error too. Let me discuss based on original code from here.

Arguments passed to _validate_elements function follows:

```python
Attribute(name='name', default=NOTHING, validator=<function type_validator.<locals>._validator at 0x7fb8d80e57a0>, repr=True, eq=True, order=True, hash=None, init=True, metadata=mappingproxy({}), type=typing.Dict[str, typing.List[str]], converter=None, kw_only=False) {'foo': []} typing.Dict[str, typing.List[str]]
Attribute(name='name', default=NOTHING, validator=<function type_validator.<locals>._validator at 0x7fb8d80e57a0>, repr=True, eq=True, order=True, hash=None, init=True, metadata=mappingproxy({}), type=typing.Dict[str, typing.List[str]], converter=None, kw_only=False) foo <class 'str'>
Attribute(name='name', default=NOTHING, validator=<function type_validator.<locals>._validator at 0x7fb8d80e57a0>, repr=True, eq=True, order=True, hash=None, init=True, metadata=mappingproxy({}), type=typing.Dict[str, typing.List[str]], converter=None, kw_only=False) [] typing.List[str]
```

Setting `empty_ok=False`, we can check the given value itself is empty(`"foo": []}` on the test case). But as for nested value check, the process never go back to _validator function's if block. Finally _validate_elements function checks whether value '[]' is list and returns None.

If we need deep empty check to nested value, we'll need make any changes to below _validator function but it might have any side effect...

Reply:
Ok, I think this is fine, if I think about it in more detail now, I'm not sure how checking for nested empty values would be anyhow useful, but we can add this in the future.